simba/schedule.py

```diff
@@ -604,6 +628,43 @@ def generate_scenario(self, args):
                             ext_load["csv_file"] = str(ext_load_path)
                             events["fixed_load"][gc_name + " ext. load"] = ext_load
 
+                        # temporary lowering of grid connector max power during peak load windows
+                        if time_windows is not None:
+                            grid_operator = station.get('grid_operator')
+                            if grid_operator is None:
+                                warnings.warn('Peak Load Windows given, '
+                                              f'but {gc_name} has no grid operator')
+                            # get relevant peak load windows
+                            windows = time_windows.get(grid_operator)
+                            if grid_operator is not None and windows is None:
+                                warnings.warn(f'Grid operator {grid_operator} '
+                                              'not in time windows')
+                            if grid_operator is not None and windows is not None:
+                                # generate grid operator events for simulation
+                                logging.debug(f"{gc_name} time windows: {windows}")
+                                reduced_power = station.get(
+                                    'peak_load_window_power',
+                                    vars(args).get('peak_load_window_power_' + station_type))
+                                assert reduced_power <= gc_power
+                                in_window = False
+                                # can't directly iterate over datetimes:
+                                # get number of simulation timesteps (round up)
+                                n_intervals = -((start_simulation - stop_simulation) // interval)
+                                for t_idx in range(n_intervals):
+                                    cur_time = start_simulation + t_idx * interval
+                                    cur_window = spice_ev_util.datetime_within_time_window(
+                                        cur_time, windows, voltage_level)
+                                    if cur_window ^ in_window:
+                                        # window change: generate event
+                                        events["grid_operator_signals"].append({
+                                            # known one day in advance
+                                            "signal_time": (cur_time - daily).isoformat(),
+                                            "start_time": cur_time.isoformat(),
+                                            "grid_connector_id": gc_name,
+                                            "max_power": reduced_power if cur_window else gc_power,
+                                        })
+                                        in_window = cur_window
```

Comment:
this reads like only the first window is read, since in_window is never set to false again after the first cur_window ^in_window is hit